### ISSUE TITLE
[5.2] Array Validation - Proper Hierarchy

### ISSUE THREAD

**Issue by cyrossignol** [laravel/framework#11078]
I'm deciding if I want to work on a pull request for the new array validation in 5.2. The current implementation in the branch doesn't seem to permit validation of each element in the top-level array by themselves. However, I'm not sure if the current limitations are intentional so I'd like to hear your thoughts.

Consider the following example:

```html
<input name="users[][name]" value="Taylor Otwell">
<input name="users[][name]" value="Cy Rossignol">
```

PHP will create an array from the above form like the following:

```php
[
    'users' => [
        0 => [
            'name' => 'Taylor Otwell',
        ],
        1 => [
            'name' => 'Cy Rossignol'
        ]
    ]
]
```

If we use the following validation rules:

```php
$rules = [
    'users.*.name' => 'required:string',   // works fine
    'users.*' => 'array|some_custom_rule', // can't do this
];
```

...then we cannot apply validation to the top level array.

This behavior results from the use of `Str::is()` to compare the rule key which transforms the terminating `*` of the second rule into a zero or more regex wildcard.

**Comment by antonybudianto** [comment]
I already made a similar [PR](https://github.com/laravel/framework/pull/9959) before, but rejected :sweat_smile: 
Let's see

**Comment by tomschlick** [comment]
```php
'users.*' => 'array|some_custom_rule', // can't do this
```

why would you want to do this if you were already checking for a field in the array? doing that would validate that the array was actually an array with that key in it... just looking for some use cases...

**Comment by cyrossignol** [comment]
Thanks, Tom. I wrote a contrived example above for simplicity, but I understand your point. I put the `array` rule in my example just to illustrate that one should expect an array, and I placed the two example rules side-by-side for contrast. A real-world application probably wouldn't contain rules like those together. 

I find the requested feature more useful when the fields in the child arrays relate to each other. 

For example, consider fields that represent a date range:

```html 
<input name="date_ranges[1][start_date]" value="2015-01-01">
<input name="date_ranges[1][end_date]" value="2015-02-02">

<input name="date_ranges[2][start_date]" value="2015-01-01">
<input name="date_ranges[2][end_date]" value="2015-02-02">
``` 

```php 
$rules = [
    'date_ranges.*' => 'date_range|in_year:2015',
];
```

Of course, this validation is possible without this feature. However, with the ability to validate the top-level array, we can write custom validation rules much more easily.
